README.md
# funvas [![Publish workflow](https://github.com/creativecreatorormaybenot/funvas/workflows/Publish/badge.svg)](https://github.com/creativecreatorormaybenot/funvas/actions) [![GitHub stars](https://img.shields.io/github/stars/creativecreatorormaybenot/funvas.svg)](https://github.com/creativecreatorormaybenot/funvas) [![Pub version](https://img.shields.io/pub/v/funvas.svg)](https://pub.dev/packages/funvas) [![Twitter Follow](https://img.shields.io/twitter/follow/creativemaybeno?label=Follow&style=social)](https://twitter.com/creativemaybeno)

Flutter package that allows creating canvas animations based on time and math functions.

The name "funvas" is based on Flutter + fun + canvas. Let me know if you have any better ideas :)

## Concept

The idea of the package is to provide an easy way to create custom canvas animations based only
on time and some math functions (sine, cosine, etc.) - like [this one][Twitter].

<p float="left">
<a target="_blank" href="https://twitter.com/creativemaybeno/status/1328261273922973696?s=20"><img src="https://s8.gifyu.com/images/animation8709ccbbf7b20e6e.gif" width="49%"></a>
<a target="_blank" href="https://twitter.com/creativemaybeno/status/1327309901270560769?s=20"><img src="https://s8.gifyu.com/images/animation8709ccbbf7b20e6f.gif" width="49%"></a>
</p>

*Inspired by Dwitter* ([check it out][Dwitter]). This is also the reason why
the following shortcut functions and variables are available; they might be expanded upon in the 
future given that there are a lot more possibilities:

```text
u(t) is called 60 times per second.
    t: Elapsed time in seconds.
    S: Shorthand for sin from dart:math.
    C: Shorthand for cos from dart:math.
    T: Shorthand for tan from dart:math.
    R: Shorthand for Color.fromRGBA, usage ex.: R(255, 255, 255, 0.5)
    c: A dart:ui canvas.
    x: A context for the canvas, providing size.
```

You can of course use all of the `Canvas` functionality, the same way you can use them in a
`CustomPainter`; the above is just in homage to Dwitter :)

## Usage

You create funvas animations by extending `Funvas` and you can display the animations using a
[FunvasContainer].
Note that you have to size the animation from outside, e.g. using a [SizedBox].

```dart
import 'package:flutter/material.dart';
import 'package:funvas/funvas.dart';

/// Example implementation of a funvas.
///
/// The animation is drawn in [u] based on [t] in seconds.
class ExampleFunvas extends Funvas {
  @override
  void u(double t) {
    c.drawCircle(
      Offset(x.width / 2, x.height / 2),
      S(t).abs() * x.height / 4 + 42,
      Paint()..color = R(C(t) * 255, 42, 60 + T(t)),
    );
  }
}

/// Example widget that displays the [ExampleFunvas] animation.
class ExampleFunvasWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 420,
      height: 420,
      child: FunvasContainer(
        funvas: ExampleFunvas(),
      ),
    );
  }
}
```

See the [example package][example] for a complete example implementation.

## Exporting animations

In the [example package][example], there is also an exporter that can be used to export funvas
animations directly to GIF. See the `example/README.md` file for an explanation.

[Twitter]: https://twitter.com/creativemaybeno/status/1285343758247178240?s=20
[Dwitter]: https://www.dwitter.net/about 
[example]: https://github.com/creativecreatorormaybenot/funvas/tree/master/example


CHANGELOG.md
## 0.1.0-nullsafety.0

* Added breaking change version.
  Of course, this is somewhat pointless now that the breaking change is already available
  as `0.0.4`, however, we should be fine by not publishing that as a pre-release. And this is
  especially fine as the migration was meant to be a demonstration :)

## 0.0.4-nullsafety.3

* Ran `dart format` to format all files.

## 0.0.4-nullsafety.2

* Fixed changelog.

## 0.0.4-nullsafety.1

* Migrated to null safety.

## 0.0.3+1

* Removed unnecessary publish workflow badge from README.

## 0.0.3

* Included example GIFs in the README.
* Added gifsicle optimization in exporter.
* Improved package description.

## 0.0.2

* Expanded example by adding funvas animations and an exporter for exporting funvas animations to
  GIF.
* Updated README to include usage guide and mention the exporter.

## 0.0.1

* Pre-release version of funvas that is a functioning package but not yet polished.


example/lib/main.dart
import 'dart:ui';

import 'package:example/tweets/tweets.dart';
import 'package:example/viewer.dart';
import 'package:flutter/material.dart';
import 'package:funvas/funvas.dart';

void main() {
  runApp(const ExampleApp());
}

/// Example app demonstrating how to integrate a [Funvas] using a
/// [FunvasContainer].
class ExampleApp extends StatelessWidget {
  /// Constructs [ExampleApp].
  const ExampleApp({Key key}) : super(key: key);

  /// List of example funvas implementations.
  static final examples = <Funvas>[
    ExampleFunvas(),
    WaveFunvas(),
    OrbsFunvas(),
  ];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: DefaultTabController(
        length: 2,
        child: Scaffold(
          appBar: AppBar(
            title: Text('creativecreatorormaybenot Funvas'),
            bottom: TabBar(
              isScrollable: true,
              tabs: [
                Tab(text: 'Examples'),
                Tab(text: '@creativemaybeno tweets'),
              ],
            ),
          ),
          body: TabBarView(
            children: [
              FunvasViewer(funvases: ExampleApp.examples),
              FunvasViewer(funvases: creativecreatorormaybenotTweets),
            ],
          ),
        ),
      ),
    );
  }
}

/// Example implementation of a funvas.
///
/// It uses a very simple [Canvas.drawCircle] operation to create an animation.
class ExampleFunvas extends Funvas {
  @override
  void u(double t) {
    c.drawCircle(
      Offset(x.width / 2, x.height / 2),
      S(t).abs() * x.height / 4 + 42,
      Paint()..color = R(C(t) * 255, 42, 60 + T(t)),
    );
  }
}

/// Funvas adapted 1:1 from https://www.dwitter.net/d/3713.
class WaveFunvas extends Funvas {
  @override
  void u(double t) {
    c.scale(x.width / 1920, x.height / 1080);

    for (var i = 0; i < 64; i++) {
      c.drawRect(
        Rect.fromLTWH(
          i * 30.0,
          400 + C(4 * t + (i * 3)) * 100,
          27,
          200,
        ),
        Paint(),
      );
    }
  }
}

/// Funvas adapted 1:1 from https://www.dwitter.net/d/4342.
class OrbsFunvas extends Funvas {
  @override
  void u(double t) {
    c.scale(x.width / 1920, x.height / 1080);

    final v = t + 400;
    for (var q = 255; q > 0; q--) {
      final paint = Paint()..color = R(q, q, q);
      c.drawCircle(
          Offset(
            1920 / 2 + C(v - q) * (v + q),
            540 + S(v - q) * (v - q),
          ),
          40,
          paint);
    }
  }
}
